### ISSUE TITLE
Pretty page formatter

### ISSUE THREAD

**Issue by sagikazarmark** [thephpleague/booboo#5]
I am not sure what the plans are, but this package could borrow the pretty page handler from whoops which is awesome (at least tiny).

My favourite feature is the open in editor, which makes it super easy to debug the application and solve problems fast.

**Comment by brandonsavage** [comment]
I was planning to borrow heavily from theirs, and make improvements, too. I'm going to probably target that for a future release (maybe 1.1?). I don't think it needs a major release, since we're adding new features to the API but not removing any (check my logic, @philsturgeon?)

**Comment by sagikazarmark** [comment]
Actually I am thinking about if it should be in a separate package or not. It adds a lot of resources (views, assets) which should not be in the core package. Maybe `booboo-pretty`?

I have some improvement ideas as well, I would be happy to provide a PR if you are fine with that.

**Comment by brandonsavage** [comment]
A separate package actually makes a lot of sense.

PRs are always welcome! I recommend breaking work into small, discrete
chunks, in case something won't fly or requires more discussion than
others. I hate turning something down that has lots of good stuff and one
thing I don't care for.

**Comment by sagikazarmark** [comment]
My ideas:

* Obviously implement it as a BooBoo handler
* We should provide a `TemplateEnginge` interface which can be implemented by any frameworks view/display package, also with a default implementation
* I would use assetic or some asset manager to handle assets (which gives more control to the user how it should be done: cache assets, always load them, etc)

I am thinking about starting it as a separate package instead of a PR, so we can always change things and I can hand it over any time you think it is finished. What do you think?

**Comment by sagikazarmark** [comment]
Is there a reason behind having the Whoops not completely ported? (I am referring to `Frame::getFileLines` and `Frame::getFileContents`)

**Comment by brandonsavage** [comment]
Since I wasn't implementing them, I removed them. I didn't want methods that were either not required by the interface being used or not required by the component itself. Feel free to add them back if you need them.

**Comment by sagikazarmark** [comment]
Will do.

**Comment by sagikazarmark** [comment]
Here is an initial implementation:

https://github.com/sagikazarmark/booboo-pretty

I could not implement open editor feature so far, and I am not really satisfied with the template/asset handling either, but this is what I could come up with so far. I am open to suggestions.

(I will also do some asset refactoring, for example better formatting of CSS, minification, probably a less implementation for development purposes, and using local copy of cdn resources)

**Comment by sagikazarmark** [comment]
I updated the code with the following features:

* I created a less version of the style which is easier to edit/extend (it always have to be parsed upon change, I added a robo task for it)
* Assets are minified
* I try to make everything local, but it seems the ZeroClipboard swf must be either placed into a public path or retrieved from the internet (no "offline" usage)
* Template and Asset helpers are merged, because framework integrations probably have their own way to handle assets.
* I added `Table`s instead of simple array data tables. This way dynamic data can be passed to the formatter, for example a callable or any kind of object which needs to extract data before printing it.

**Comment by sagikazarmark** [comment]
@brandonsavage I opened a few issues in the created repository, I would be happy to hear your opinion on those as well.

**Comment by brandonsavage** [comment]
I'm going to try and find some time to look at this during the week. Thank
you for your work!

**Comment by sagikazarmark** [comment]
Great, let me know if you have something on your mind.

**Comment by sagikazarmark** [comment]
@brandonsavage any chance you can look into this in the next few days? I would like to replace whoops with booboo in FuelPHP, but we need a pretty page formatter thing.

**Comment by jaoued** [comment]
+1

**Comment by brandonsavage** [comment]
I still need to take a look at this, I've been on travel but it's on my
list.

**Comment by sagikazarmark** [comment]
@brandonsavage ping

**Comment by sagikazarmark** [comment]
@brandonsavage Is there a chance you can look into this anytime soon? Not just because I would like to make it as the "official" pretty page handler, but because I would like to hear your opinion as well.

**Comment by brandonsavage** [comment]
@sagikazarmark First, my deepest apologies for not getting to this sooner. Really, I am sorry.

This looks like a great first step. I'd like to see "no stack" when there's no call stack to show (right now it's just a blank whitespace). Obviously it also needs tests. :) :+1:

**Comment by sagikazarmark** [comment]
Yeah, I am also concerned about tests. However (after a long time) the implementation seems to me a bit overcomplicated. Basically it is the same as in case of whoops with a little tweaking. But I am not sure we should provide anything like that finder logic.

It's obvious that different frameworks use different templating engines, views which should be supported. But I wonder if asset finding should also be implemented by frameworks instead of having this DefaultTemplate helper thing.

So my plan is to make things way much simpler: remove the finder thing entirely. Replace the current template helper implementation with one that has the paths to the bundled assets and templates so that framework integrations MUST implement their own logic for that.

One more thing I am concerned about is the name of the main layout file. Currently it's burned into the code: https://github.com/sagikazarmark/booboo-pretty/blob/master/src/Formatter/PrettyPageFormatter.php#L76 So my plan for the template helper is to have the main render function accept only a data parameter (or accept parameters directly?), because the only reason the template name is there is that it is used in the default template helper?

tl;dr: the general way I would like to follow is making the whole thing simpler.

What do you think? Anything to add?

**Comment by sagikazarmark** [comment]
By no stack trace you mean the case when there are no frames?

https://github.com/sagikazarmark/booboo-pretty/blob/master/src/Formatter/PrettyPageFormatter.php#L176

**Comment by brandonsavage** [comment]
Yes, when there are no frames, we ought to tell the user something.

**Comment by sagikazarmark** [comment]
I also wonder how we should handle this case:

https://github.com/sagikazarmark/booboo-pretty/blob/master/src/Formatter/PrettyPageFormatter.php#L151

**Comment by sagikazarmark** [comment]
Thinking about package architecture: it's obvious that we need a "default" implementation which just works. But in some cases, if we override the template anyway (for example when using a templating engine) it does not make sense to always include all these things.

So my plan is:
booboo-pretty: the main package containing interfaces, table implementations, assets
booboo-pretty-php: a default implementation with php templates
booboo-pretty-twig: Guess what comes here... :)

Any thoughts? Is it an overcomplication?

**Comment by brandonsavage** [comment]
I think this is a bit of an overcomplication. If we ship this formatter as
a stand-alone separate component that can be installed or removed at will
(and not part of the core package), we avoid many of the architectural
decisions that might otherwise be required. I also think throwing Twig into
the mess would just be way too much.

**Comment by sagikazarmark** [comment]
One thing I am not sure what extension points should we provide then?

Assets? Obviously. I think the ideal would be to recommend copying the assets into the public directory of the webserver.

Templates?

TemplateHelpers to allow custom templating?

**Comment by brandonsavage** [comment]
This formatter should never be displayed in production. Thus I'm okay with
a hack that lets us display things from within the vendor directory.

**Comment by sagikazarmark** [comment]
My biggest problem with that is the Copy function needs an swf file, which cannot be simply inserted into the HTML code. Currently (and in whoops as well) it is linked from the internet (CDN) which is not a really nice solution.

Anyway, I don't have a problem with that.

**Comment by mindplay-dk** [comment]
Just my two cents, but I think you should integrate pretty printing as an optional feature in the main package - nobody wants to search for additional packages or install a bunch of them to get a full, working, useful, comfortable error-handler. I think part of the reason Whoops is popular (200k installs/month) is because it has everything straight out-of-the-box, ready to run. Just my opinion.

**Comment by sagikazarmark** [comment]
@mindplay-dk I would agree with you. But this pretty printing solution is rather ugly and it should not be used in production. That's why we agreed with @brandonsavage to separate it. It should be required maximum as a development dependency.

And to be fair: Whoops has really high download count because it is the default error handler in laravel: :stuck_out_tongue_winking_eye:

**Comment by sagikazarmark** [comment]
@brandonsavage what's the situation with this?

**Comment by sagikazarmark** [comment]
I am thinking about replacing ZeroClipboard with [clipboard.js](https://zenorocha.github.io/clipboard.js/) So that we can get rid of external dependencies.

Also, after a long thinking (and gathering experience in packaging with [httplug](http://httplug.io), I tend to agree with @mindplay-dk now: it would be better to have this in the core package to reduce dependencies and reduce the hassle to set up dependencies.

**Comment by BafS** [comment]
Hi,

I have begin to develop a "pretty" formatter. It is far from finished but Booboo handle errors in a cleaner way than Woops, it just missed the pretty formatter imho.

My fork: https://github.com/BafS/booboo

![screen shot 2016-04-06 at 23 16 16](https://cloud.githubusercontent.com/assets/588205/14333656/2725cf64-fc50-11e5-8719-5797d9bb8c02.png)

**Comment by mindplay-dk** [comment]
Looks nice :-)

The top item in "stack frames" is not actually a stack frame though - it's an "uncaught exception", the stack trace is everything below it.

I guess this is an almost standard misunderstanding in PHP land though - displaying the error site as though it were part of the stack trace. Both symfony and php-unit do that.

**Comment by brandonsavage** [comment]
Closing since work seems to have stopped here; will reopen if work resumes.

**Comment by mzdr** [comment]
Hey guys!

Just for your information… I stumbled upon this issue, read the whole conversation, checked out the existing solutions and finally got pretty interested in creating my own pretty page formatter. 

It’s far from being finished, doesn’t have any fancy frontend features (yet) but should serve a basic set of needs. You can check it out over here: [mzdr/oh-snap](https://github.com/mzdr/oh-snap)

![Screenshot](https://github.com/mzdr/oh-snap/raw/master/screenshot.jpg)

Feel free to contribute, create issues or drop a line with your feedback. Thanks for making BooBoo. 👍
